lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp

```diff
@@ -257,40 +272,73 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
          "endBB is required to post-dominate startBB");
   SmallVector<std::unique_ptr<SESERegionTree>, 4> results;
 
-  // Iteratively work our way up the post-dominator tree (moving startBB until
+  // Iteratively work our way up the post-dominator tree (moving currentBB until
   // we reach the endBB), producing a sequence of loops, diamonds, and single
   // block nodes as we go.
-  while (startBB != endBB) {
+  SILBasicBlock *currentBB = startBB;
+  while (currentBB != endBB) {
+    // If currentBB is not dominated by startBB, we create a function region.
+    if (!DI.dominates(startBB, currentBB)) {
+      auto iter = functionRegions.find(currentBB);
+      if (iter == functionRegions.end()) {
+        // Create and cache the function region.
+        llvm::SmallVector<SILBasicBlock *, 32> descendants;
+        DI.getDescendants(currentBB, descendants);
+        SILBasicBlock *subRegionEndBB = startBB;
+        for (SILBasicBlock *otherBB : descendants) {
+          subRegionEndBB = PDI.findNearestCommonDominator(subRegionEndBB, otherBB);
+        }
+        // We need to create a new SESE Region so that this will end up as a
+        // separate function in the lowered graph.
+        std::shared_ptr<SESERegionTree> subSESERegion(
+            processAcyclicRegion(currentBB, subRegionEndBB).release());
+        SILBasicBlock *nextBB = nullptr;
+        if (subRegionEndBB == endBB) {
+          nextBB = endBB;
+        } else {
+          auto *branch = dyn_cast<BranchInst>(subRegionEndBB->getTerminator());
+          assert(branch &&
+                 "Function region should end with an unconditional branch");
+          nextBB = branch->getDestBB();
+        }
+        auto emplace_result = functionRegions.try_emplace(
+            currentBB, std::make_pair(subSESERegion, nextBB));
+        iter = emplace_result.first;
+      }
+      results.push_back(llvm::make_unique<FunctionSESERegion>(iter->second.first));
+      currentBB = iter->second.second;
```

Comment:
IIUC, in this code path where the function region has been processed before, and we are reading from the existing entry in `functionRegions`, `currentBB` will be set to bb7 in the example in PR description, and `bb4` is the function region.

In that case, would we be lowering `bb7` twice, before it has been lowered the first time we are processing `bb4`, and in this code path we will be lowering it again?

Reply:
in the unit test, the "exit block" of the shared region happens to be endBB. But in a more complex test case, could a shared region exit to other BBs, and not endBB?